google/cloud/bigtable/batcher.py

```diff
@@ -125,22 +236,86 @@ def mutate_rows(self, rows):
                    match the number of rows that were retried
                  * :exc:`.batcher.MaxMutationsError` if any row exceeds max
                    mutations count.
+                 * :exc:`.batcherMutationsBatchError` if there's any error in the
+                   mutations.
         """
         for row in rows:
             self.mutate(row)
 
     def flush(self):
-        """Sends the current. batch to Cloud Bigtable.
+        """Sends the current batch to Cloud Bigtable synchronously.
         For example:
 
         .. literalinclude:: snippets.py
             :start-after: [START bigtable_api_batcher_flush]
             :end-before: [END bigtable_api_batcher_flush]
             :dedent: 4
 
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+        rows_to_flush = []
+        while not self._rows.empty():
+            rows_to_flush.append(self._rows.get())
+        response = self.flush_rows(rows_to_flush)
+        return response
+
+    def flush_async(self):
+        """Sends the current batch to Cloud Bigtable asynchronously.
+
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+
+        rows_to_flush = []
+        while not self._rows.empty():
+            rows_to_flush.append(self._rows.get())
+        future = self._executor.submit(self.flush_rows, rows_to_flush)
+        # catch the exceptions in the mutation
+        exc = future.exception()
+        if exc:
+            raise exc
+        else:
+            result = future.result()
+        return result
+
+    def flush_rows(self, rows_to_flush=None):
+        """Mutate the specified rows.
+
+        raises:
+            * :exc:`.batcherMutationsBatchError` if there's any error in the
+                mutations.
+        """
+        response = []
+        if len(rows_to_flush) > 0:
+            # returns a list of status codes
+            response = self.table.mutate_rows(rows_to_flush)
+            responses = []
+            has_error = False
+            for result in response:
+                if result.code != 0:
+                    has_error = True
+                responses.append(result)
+
+            if has_error:
+                raise MutationsBatchError(status_codes=response)
```

Comment:
I noticed Status codes can be converted into exceptions using [google.api_core.exceptions.from_grpc_status](https://github.com/googleapis/python-api-core/blob/b6eb6dee2762b602d553ace510808afd67af461d/google/api_core/exceptions.py#L441). Maybe it would be better to raise those directly, to give the full context, rather than using the raw status codes?